### PR TITLE
Improve testability of Ninja applications (programmatic configuration of guice bindings and ninjaProperties)

### PR DESCRIPTION
**Pull request by raphaelbauer**
This PR adds two massive features to Ninja:
- It allows to programmatically override properties from your tests.
- It allows to programmatically override guice binding from your tests.

TODO:
 - [x] documentation
 - [x] tests

### CODE REVIEW

**Comment by jjlauer** [comment]
This is super cool.  Love it.

File: ninja-core/src/main/java/ninja/standalone/AbstractConsole.java [hunk 1]
```diff
@@ -35,13 +38,16 @@
...
     // can all be changed prior to configure()
     protected NinjaMode ninjaMode;
-    protected String externalConfigurationPath;
+    protected Optional<String> externalConfigurationPath = Optional.empty();
     protected String name;
     // internal state
     protected boolean configured;
     protected boolean started;
     protected NinjaPropertiesImpl ninjaProperties; // after configure()
     protected OverlayedNinjaProperties overlayedNinjaProperties; // after configure()
```
Comment (raphaelbauer) [inline ninja-core/src/main/java/ninja/standalone/AbstractConsole.java:47]: I have actually not really an idea why we have OverlayedNinjaProperties...

File: ninja-core/src/site/markdown/developer/getting_started.md [hunk 2]
```diff
@@ -33,6 +33,19 @@ send us a pull request to become a happy member of the Ninja family :)
 
 ### Code style
 
+- **Don't use null**
```
Comment (raphaelbauer) [inline ninja-core/src/site/markdown/developer/getting_started.md:36]: @jjlauer Does that makes sense to you? I added some guidelines...

File: ninja-core/src/test/java/ninja/utils/NinjaPropertiesImplTest.java [hunk 3]
```diff
@@ -43,24 +41,19 @@ public void tearDown() {
     public void testSkippingThroughModesWorks() {
 
         // check that mode tests works:
-        NinjaPropertiesImpl ninjaPropertiesImpl = new NinjaPropertiesImpl(NinjaMode.test);
```
Comment (raphaelbauer) [inline ninja-core/src/test/java/ninja/utils/NinjaPropertiesImplTest.java:46]: Nothing hass changed. I just moved from hamcrest to AssertJ + added one test at the end...

**Comment by raphaelbauer** [comment]
@jjlauer Would you mind having a look? I know I am a bad person - and I should have made two PRs I guess. Anyhow...
All feedback from your side would be great. Thanks!

These were two features that were missing deeply from Ninja. Now we got them. It puts also Ninja on-par (or ahead) with the best Java frameworks in terms of testability afaics.

File: ninja-core/src/main/java/ninja/standalone/AbstractConsole.java (hunk 1 above)
Comment (jjlauer) [reply ninja-core/src/main/java/ninja/standalone/AbstractConsole.java:47]: I remember implementing it when working on the new standalone class hierarchy.  Your new override stuff may make it obsolete, but we'd have to dive in to the code to be sure.

File: ninja-core/src/site/markdown/developer/getting_started.md (hunk 2 above)
Comment (jjlauer) [reply ninja-core/src/site/markdown/developer/getting_started.md:36]: These are good.  I haven't personally switched to AssertJ myself, but I checked them out and they look interesting.  I am happy to give them a shot.

**Comment by jjlauer** [comment]
Reviewed everything, looks good IMHO.

**Comment by raphaelbauer** [comment]
It's live. And I am using it already in one of my apps. Whoop whoop.

**Pull request merged**